activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -76,7 +76,9 @@ def replace(record)
         def update_counters(by)
           if require_counter_update? && foreign_key_present?
             if target && !stale_target?
-              target.increment!(reflection.counter_cache_column, by, touch: reflection.options[:touch])
+              target.increment!(reflection.counter_cache_column, by, touch: reflection.options[:touch]) &&
+                (target[target.class.locking_column] += 1 if target.locking_enabled?)
```

Comment:
This will leave the record dirty, i.e. `lock_version_changed?` will return true, and saving it will write to the database again. That could be solved with `clear_attribute_change`, but it seems like the root of the problem is actually the `increment!` method, since it's reproducible by using it directly:

```ruby
car = Car.create!
car.increment!(:wheels_count)
car.update!(name: "herbie")
```

Reply:
Yeah, it should be solved in `increment!`.

Reply:
Thank you @eugeneius and @kamipo. I've moved the patch to `ActiveRecord::Persistence#increment!` and called `clear_attribute_change` after modifying an instance's lock version. (Also added an assertion that lock_version doesn't change.)